common/crypto/attestation-api/evidence/verify-evidence.cpp

```diff
@@ -0,0 +1,88 @@
+/*
+ * Copyright 2020 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "verify-evidence.h"
+#include <string.h>
+#include <string>
+#include "attestation_tags.h"
+#include "error.h"
+#include "logging.h"
+#include "types/types.h"
+#include "verify-ias-evidence.h"
+#include "verify-dcap-evidence.h"
+#include "verify-dcap-direct-evidence.h"
+
+// < JSON include
+#define JSON_HAS_CPP_11
+#define JSON_NO_IO 1
+#include <errno.h>
+#include "nlohmann/json.hpp"
+using json = nlohmann::json;
+// JSON include >
+
+bool verify_evidence(uint8_t* evidence,
+    uint32_t evidence_length,
+    uint8_t* expected_statement,
+    uint32_t expected_statement_length,
+    uint8_t* expected_code_id,
+    uint32_t expected_code_id_length)
+{
+    bool ret = false;
+    json root;
+    std::string attestation_type;
+    std::string evidence_field;
+    std::string evidence_str((char*)evidence, evidence_length);
+    ByteArray ba_expected_statement(
+        expected_statement, expected_statement + expected_statement_length);
+    ByteArray ba_expected_code_id(expected_code_id, expected_code_id + expected_code_id_length);
+
+    CATCH(ret, root = json::parse(evidence_str));
+    COND2LOGERR(!ret, "invalid evidence json");
+
+    CATCH(ret, attestation_type = root[ATTESTATION_TYPE_TAG].template get<std::string>());
+    COND2LOGERR(!ret, "invalid (or missing) attestation type field");
+
+    CATCH(ret, evidence_field = root[EVIDENCE_TAG].template get<std::string>());
+    COND2LOGERR(!ret, "invalid evidence field");
+
+    if (0 == attestation_type.compare(SIMULATED_TYPE_TAG))
+    {
+        // nothing to check
+        ret = true;
+    }
+
+    if (0 == attestation_type.compare(EPID_LINKABLE_TYPE_TAG) ||
+        0 == attestation_type.compare(EPID_UNLINKABLE_TYPE_TAG))
+    {
+        ByteArray ba_evidence(evidence_field.begin(), evidence_field.end());
+        bool b = verify_ias_evidence(ba_evidence, ba_expected_statement, ba_expected_code_id);
```

Comment:
Not sure i understand 100%.  Do we not care in PDO what attestation type is used?  Or put differentlly, the only thing we care is that there is _some_ valid attestation, regardless whether it is linkable EPID, unlinkable EPID or DCAP?   I thought previously it was crucial that we had linkable EPID?

Reply:
Oh, now I got your question.
There are two ways: 
1) as I show in the [next PR](https://github.com/hyperledger-labs/private-data-objects/pull/502/commits/77a17c7300fa6b39bdd531bb7ab7b24d58d76a61#diff-0374b424e40097c33e75b092fa06140de8704ebe9b829f80c1a842ab3bc20a89) (see signup.cpp line 141), we can force the contract enclave to generate (or to check that it is generating) a linkable EPID quote, which will be reflected in the MRENCLAVE. So, later, if you have the right mrenclave, the verification is correct.
2) the second way is to check for the the "epidpseudonym" field in the report -- this is optional and only present for linkable signatures.

We could:
* either rely on (1),
* or add the epidpseudonym check at verification time,
* or simply leave this epidpseudonym check to the policy appraisal (which is not implemented here, and should also include the debug-disabled check among others).

Reply:
> Oh, now I got your question. There are two ways:
> 
> 1. as I show in the [next PR](https://github.com/hyperledger-labs/private-data-objects/pull/502/commits/77a17c7300fa6b39bdd531bb7ab7b24d58d76a61#diff-0374b424e40097c33e75b092fa06140de8704ebe9b829f80c1a842ab3bc20a89) (see signup.cpp line 141), we can force the contract enclave to generate (or to check that it is generating) a linkable EPID quote, which will be reflected in the MRENCLAVE. 

This is code outside of enclave: why would that result in different MRENCLAVE depending on attestation type?

> So, later, if you have the right mrenclave, the verification is correct.
> 2. the second way is to check for the the "epidpseudonym" field in the report -- this is optional and only present for linkable signatures.
> 
> We could:
> 
> * either rely on (1),
> * or add the epidpseudonym check at verification time,
> * or simply leave this epidpseudonym check to the policy appraisal (which is not implemented here, and should also include the debug-disabled check among others).

My comment was originally just starting from an API perspective and my expectation would have been that there would be an a boolean flag `linkable`.  But you bring up epid-pseudonym of course is a good point.  but as pseudonym is apriori not known, my api proposal would have been to add the boolean flag as input and have an optional outparam which for linkable would return the pseudonym.   That said, at this point with EPID not really worth changing (but then my designer head cannot resist and i was actually also looking at  options for PQC-equivlants of EPID (or at least DAA). There actually would be some options and they would satisfy the [IETF IAB](https://datatracker.ietf.org/doc/statement-iab-statement-on-the-risks-of-attestation-of-software-and-hardware-on-the-open-internet/) but alas very unlikely that intel would ever pick that up unless NIST would ....

Reply:
> This is code outside of enclave: why would that result in different MRENCLAVE depending on attestation type?

You're right that the code I linked is outside of the enclave.
However, the attestation generation (get_attestation) is inside the enclave.
So, what I meant is to have a check right there to enforce that the attestation being generated is of type `epid-linkable`, as defined in the [usage ](https://github.com/hyperledger-labs/private-data-objects/pull/501/files#diff-5659628547c8fcb187e090ba6a1f5e13dca2823243fe11b4ea58cc6060ccb7c6) document.

> but as pseudonym is apriori not known

Right, but the "field" is in the report and signed by IAS. So you just have to check that it is present in this case.
And, again, this can be one of the conditions in a policy appraisal -- which should be implemented anyway.